### PR TITLE
feat(fit): crash-safe activity recording — flush + next-boot recovery + save-status (audit #4 SDK half)

### PR DESCRIPTION
**Pull request by rryles**
## Summary

Makes `.fit` activity recording crash-safe. A reset mid-recording (the field-reboot class) used to leave the `.fit` with a placeholder header (`dataSize=0`, no CRC) — unrecoverable and unimportable — so the whole workout was lost. This is the **SDK half of robustness-audit finding #4** (the kernel half — atomic settings + boot-time mount recovery — is a separate kernel PR).

Two findings:

### #11 — Flush during recording + next-boot recovery
- **`FitWriter`**: `finish()`'s back-patch / read-back-CRC / append tail is refactored into a shared `finalize(IFile&, dataEnd)` and exposed as `static recover(IFile&, dataEnd)`. `recover()` reads the on-disk header (protocol/profile survive a reboot), back-patches `dataSize`, recomputes the file CRC over `[0,dataEnd)`, writes the CRC and **truncates to `dataEnd+2`** to trim any torn tail. Idempotent — a header with a non-zero `dataSize` is treated as already finalized and left intact, so a stale recovery offset can never discard a completed session/activity. `finish()`'s normal path is byte-identical (verified by the existing encoder tests).
- **New `SDK::Fit::RecordingMarker`** (all shared, zero per-app logic): a crash marker at `<activityDir>/.recording` holding the `.fit` path + the last flushed, record-complete offset. Written after the header+definitions, refreshed on each durable flush, removed on clean `stop()`/`discard()`. `recover()` reads it, finalizes the `.fit` at that offset, and always clears the marker. **The marker itself is crash-atomic** (staged `.tmp` then rotate, with a `.bak` fallback) so a crash mid-update degrades to the previous flushed offset rather than losing the activity — mirrors the kernel's `SettingsManagerBase` rotate.
- **Per-app `ActivityWriter`** (all 6 apps): flush the `.fit` + refresh the marker every ~30 s of records and on every lap, **only when `flush()` succeeds** (so the marker never advances past non-durable data); the start-time marker is written only when `begin()` succeeded. `recoverInterrupted()` is a one-call wrapper over the shared marker recover.
- **Per-app `Service`**: `run()` finalizes any interrupted activity at startup (before the first `startTrack`) and registers it via `notifyNewActivity()`. Recovery loses at most ~30 s and produces a strictly valid, importable `.fit`.

### #40 — Propagate the save status
- `ActivityWriter::stop()` now returns `bool`, threading `ok()` / `finish()` / flush / close / `saveSummary`; `LOG_ERROR` on failure. The marker is removed **only** once the `.fit` is durably finished, so a failed save is retried by next-boot recovery.
- `Service` gates registration: `notifyNewActivity()` only on `stop()==true`, else `LOG_ERROR("activity save failed")`. No user-facing UI (per agreed scope).

## Minimal per-app footprint
All reusable logic (marker I/O + recover orchestration + FIT finalize/recover) lives in the shared Fit lib. Each app gets thin wiring only — no duplicated marker/recover logic across the 6 copies.

## Testing
- **Host tests: 152/152 pass.**
  - `FitWriter`: crash→reboot→recover, torn-tail trim, idempotent, already-finalized-with-trailing-bytes no-op, bad input.
  - `RecordingMarker`: round-trip, **crash-during-update `.bak` fallback**, cleanup of primary+bak+tmp.
  - Running `ActivityWriter`: flush cadence (30 s + lap), end-to-end recovery through the app, `stop()` success/failure gating.
- **`Cycling.uapp` arm-built** (arm-none-eabi + Ninja) to verify the per-app wiring links on target — this also caught the missing `RecordingMarker.cpp` in `cmake/una-sdk.cmake` (now fixed; host-test CMake updated too).

## Known limitations (by design)
- Recovery finalizes the `.fit` only; the sibling `.json` summary is not reconstructed. Activities register off the `.fit` path, not the `.json`, so the recovered workout still appears (its on-watch summary card may be blank until synced; the real data is in the `.fit`).
- **HRMonitor** has no activity-notify mechanism (it never has — it relies on the kernel's boot-time activity scan for every `.fit`). So a recovered HRMonitor session registers on the *next* boot's scan rather than immediately. Documented in its `run()`; not a regression.

## On-device verification (not covered by host tests / arm-link)
- [ ] Force a reset mid-activity (all 6 apps); confirm on next boot the interrupted `.fit` is finalized, imports cleanly, and the marker is gone.
- [ ] Confirm a normal stop still produces a byte-valid `.fit` and its `.json`, and registers exactly once (no double-registration vs. recovery).
- [ ] Confirm the ~30 s flush cadence has no visible recording-thread stall on eMMC.

## Summary by CodeRabbit

* **New Features**
  * Added crash/power-loss recovery for interrupted activities, with automatic recovery on app/service startup.
  * Introduced a dedicated on-disk recording marker to finalize partially written FIT recordings.

* **Bug Fixes**
  * Improved durability: recovery/marker progression now advances only after successful flushes.
  * FIT finalization is now result-driven: notifications depend on successful FIT durability, not best-effort sidecar output.
  * Discarding/failing finalization now preserves recovery for the next boot.

* **Tests**
  * Expanded host tests for marker behavior, FIT recovery edge cases, flush cadence, and recovery/failure semantics.

### CODE REVIEW

**Comment by rryles** [comment]
Thanks @coderabbitai — addressed in `dc81b823`. One correction on the "Major" orphaned-activity finding, then what I changed.

## The orphaned-`.fit` premise

The finding assumes that if `saveSummary()` fails after the FIT is finalized, the marker is already gone and *"the completed activity never gets registered"* — because there's no boot-time `.fit` scan in this repo. That's correct about the SDK repo, but the registration mechanism lives in the **kernel** (separate repo), and it does auto-register — just not via a scan:

- `FileGuard::close()` fires `notifyFileAppeared` for any file that was written (`FileSystemGuard.cpp:296-305`).
- → `Application::onFileAppeared()` filters to `/Activity/*.fit` and calls `mActivityFileRegistry.onFileAppeared()`, which **adds it to the persisted, mobile-synced registry**, and sets `mHasNewActivity` (`Application.cpp:267-278`).
- `mHasNewActivity` is flushed as `onNewActivity` when the app stops (`Application.cpp:152-156`); the explicit `REQUEST_APP_NEW_ACTIVITY` (`:1293`) is just the *immediate* variant.

So a durably-finalized `.fit` is registered the moment its guarded handle closes — during a normal stop **and** during recovery's finalize-write — and `recoverInterrupted()` additionally sends the explicit notify. It is never orphaned by a summary failure. Keeping the marker until the summary succeeds wouldn't help either: recovery only finalizes the FIT (idempotent), it can't rebuild a `.json`, so the marker's sole job is FIT durability.

## What I did change

The underlying smell is real: `stop()` was conflating FIT-durability with summary-success in one bool, so a summary-only failure mislabeled the whole save as failed and suppressed the immediate notify. Fixed by **decoupling** them (all 5 summary-writing apps; HRMonitor has no summary):

- `stop()` now returns **FIT durability only**. The `.json` summary is best-effort — attempted, `LOG_ERROR` on failure, but it no longer folds into the return value or gates `notifyNewActivity()`. `mMarker.remove()` stays gated on FIT durability (unchanged).
- New test `StopSucceedsWhenOnlySummaryFails`: FIT finished/closed but the `.json` open fails → `stop()` returns true, marker cleared, `.fit` CRC-valid.

## Nitpicks

- **`recover()` write-open on bad input** — valid, fixed. The initial inspection is now a **read-only** open (`open(wMode=false)`); write access happens only inside `finalize()` after the `dataSize==0` crash-marker check, so `recover()` on a missing/invalid path no longer risks creating a stray 0-byte `.fit`. New test `RecoverOnMissingFileDoesNotCreateIt`.
- **Isolated FIT-ok/summary-fail test** — added (see above).

Host tests green (incl. the 2 new); Running `.uapp` arm-built to confirm the per-app changes link on target.

**Comment by rryles** [comment]
@coderabbitai review

File: Libs/Source/Fit/FitWriter.cpp [hunk 1]
```diff
@@ -136,66 +136,179 @@ bool FitWriter::emitData(uint8_t localType, const std::vector<uint8_t>& payload)
...
+            && hdr[8] == '.' && hdr[9] == 'F' && hdr[10] == 'I' && hdr[11] == 'T') {
+            const uint32_t dataSize =
+                static_cast<uint32_t>(hdr[4]) | (static_cast<uint32_t>(hdr[5]) << 8)
+                | (static_cast<uint32_t>(hdr[6]) << 16)
+                | (static_cast<uint32_t>(hdr[7]) << 24);
+            if (dataSize != 0) {
+                file.close();
+                return true;  // already finalized -- do not re-finalize
+            }
+        }
```
[the remark this answers is not in the document]
Comment (rryles) [reply Libs/Source/Fit/FitWriter.cpp]: Good catch — fixed in fb52dea0. Removed the `dataSize != 0 → return true` early-return. `recover()` now derives the effective data-end from the header's own `dataSize` (falling back to the marker offset when `dataSize == 0`, or when the header over-claims vs the actual file size) and **always** runs the idempotent `finalize()`. This repairs the exact window you describe (header flushed, CRC missing/torn → CRC rewritten), while still not re-finalizing at the stale marker offset (the earlier round's concern), since it trusts the header's `dataSize`. New test `RecoverRepairsCrashMidFinalize`; the trailing-bytes test now asserts the tail is trimmed rather than accepted.

File: Examples/Apps/Treadmill/Software/Libs/Sources/ActivityWriter.cpp [hunk 2]
```diff
@@ -116,6 +120,16 @@ void ActivityWriter::start(const AppInfo& info)
...
+
+    // Header + all definitions are on disk: flush and drop the recovery marker.
+    // getPosition() here is a clean record boundary, so a crash after this point
+    // is recoverable up to at least the header/definitions. Only write the
+    // marker when begin() succeeded: a failed begin() leaves a near-empty/broken
+    // .fit that the recovery marker must not point next boot's recover() at.
+    if (begun) {
+        mFile->flush();
+        mMarker.write(mFile->getPath(), static_cast<uint32_t>(mFile->getPosition()));
+    }
```
[the remark this answers is not in the document]
Comment (rryles) [reply Examples/Apps/Treadmill/Software/Libs/Sources/ActivityWriter.cpp]: Fixed in fb52dea0 — the start()-time marker write is now gated on `mFile->flush()` succeeding (`if (begun && mFile->flush())`), consistent with the existing addRecord()/addLap() gating. Applied identically across all 6 apps.

File: Examples/Apps/Cycling/Software/Libs/Sources/ActivityWriter.cpp [hunk 3]
```diff
@@ -265,29 +299,51 @@ void ActivityWriter::stop(const TrackData& track)
...
+    // FIT durability IS the save-success contract: the kernel auto-registers the
+    // .fit the moment its FileGuard::close() fires (and recoverInterrupted()
+    // re-registers after a crash), so once `ok` is true the activity is never
+    // orphaned. The .json summary is auxiliary/best-effort — recovery cannot
+    // rebuild it — so a summary-only failure must NOT suppress registration.
+    // Attempt it, log on failure, but gate the return value on FIT durability.
+    if (!saveSummary(track)) {
+        LOG_ERROR("Activity summary (.json) save failed; FIT is durable and registered\n");
+    }
+    return ok;
```
[the remark this answers is not in the document]
Comment (rryles) [reply Examples/Apps/Cycling/Software/Libs/Sources/ActivityWriter.cpp]: Fixed in fb52dea0 — `saveSummary()` is now gated on FIT durability (`if (ok && !saveSummary(track))`), so no `.json` sidecar is written for a non-durable/marker-preserved `.fit`. Applied to all 5 summary apps (Running/Cycling/Hiking/Treadmill/Workout).

File: Examples/Apps/Treadmill/Software/Libs/Sources/ActivityWriter.cpp [hunk 4]
```diff
@@ -305,29 +339,51 @@ void ActivityWriter::stop(const TrackData& track)
...
+    // FIT durability IS the save-success contract: the kernel auto-registers the
+    // .fit the moment its FileGuard::close() fires (and recoverInterrupted()
+    // re-registers after a crash), so once `ok` is true the activity is never
+    // orphaned. The .json summary is auxiliary/best-effort — recovery cannot
+    // rebuild it — so a summary-only failure must NOT suppress registration.
+    // Attempt it, log on failure, but gate the return value on FIT durability.
+    if (!saveSummary(track)) {
+        LOG_ERROR("Activity summary (.json) save failed; FIT is durable and registered\n");
+    }
+    return ok;
```
[the remark this answers is not in the document]
Comment (rryles) [reply Examples/Apps/Treadmill/Software/Libs/Sources/ActivityWriter.cpp]: Fixed in fb52dea0 — summary now gated on `ok` (`if (ok && !saveSummary(track))`). Applied across all 5 summary apps.

File: Examples/Apps/Workout/Software/Libs/Sources/ActivityWriter.cpp [hunk 5]
```diff
@@ -221,29 +255,51 @@ void ActivityWriter::stop(const TrackData& track)
...
+    // FIT durability IS the save-success contract: the kernel auto-registers the
+    // .fit the moment its FileGuard::close() fires (and recoverInterrupted()
+    // re-registers after a crash), so once `ok` is true the activity is never
+    // orphaned. The .json summary is auxiliary/best-effort — recovery cannot
+    // rebuild it — so a summary-only failure must NOT suppress registration.
+    // Attempt it, log on failure, but gate the return value on FIT durability.
+    if (!saveSummary(track)) {
+        LOG_ERROR("Activity summary (.json) save failed; FIT is durable and registered\n");
+    }
+    return ok;
```
[the remark this answers is not in the document]
Comment (rryles) [reply Examples/Apps/Workout/Software/Libs/Sources/ActivityWriter.cpp]: Fixed in fb52dea0 — summary now gated on `ok` (`if (ok && !saveSummary(track))`). Applied across all 5 summary apps.

File: Examples/Apps/Cycling/Software/Libs/Sources/ActivityWriter.cpp [hunk 6]
```diff
@@ -265,29 +299,43 @@ void ActivityWriter::stop(const TrackData& track)
...
 
-    saveSummary(track);
+    // The .fit is durably finished on disk: drop the recovery marker so the
+    // next boot does not treat this activity as interrupted.
+    if (ok) {
+        mMarker.remove();
+    }
+
+    const bool summaryOk = saveSummary(track);
+    return ok && summaryOk;
```
[the remark this answers is not in the document]
Comment (rryles) [reply Examples/Apps/Cycling/Software/Libs/Sources/ActivityWriter.cpp]: Superseded/outdated. The premise (a finalized `.fit` would be orphaned) doesn't hold: the kernel auto-registers any `.fit` on its guarded `FileGuard::close()` (and recovery re-notifies), so the marker's only job is FIT durability — recovery can't rebuild a `.json`. Addressed in dc81b823 by decoupling summary from durability (`stop()` returns FIT-durability; summary best-effort). Resolving.

**Review by sdvsaienko — APPROVED** [review]
